File: src/main/java/com/huffnest/io/BitWriter.java
package com.huffnest.io;

import com.huffnest.util.BitMerger;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.file.Files;
import java.nio.file.Path;

public class BitWriter {

  public BitWriter(Path path) throws IOException {
    os = Files.newOutputStream(path);
  }

  private OutputStream os;
  private BitMerger bitMerger = new BitMerger();

  public void pushBit(byte bit) throws IOException {
    bitMerger.pushBit(bit);

    if (bitMerger.hasByte()) {
      writeBuffer(bitMerger.getByte());
    }
  }

  public void pushByte(byte b) throws IOException {
    bitMerger.pushByte(b);
    writeBuffer(bitMerger.getByte());
  }

  public void pushInt(int value) throws IOException {
    pushByte((byte) (value >>> 24));
    pushByte((byte) (value >>> 16));
    pushByte((byte) (value >>> 8));
    pushByte((byte) value);
  }

  public void close() throws IOException {
    while (!bitMerger.isBuildInProgress()) bitMerger.pushBit((byte) 0);
    while (bitMerger.hasByte()) writeBuffer(bitMerger.getByte());
    os.close();
  }

  private void writeBuffer(byte b) throws IOException {
    os.write(new byte[] { b });
  }
}

File: HuffNest/IO/BitWriter.cs
using Util;

namespace IO;

public class BitWriter
{
    public BitWriter(string path)
    {
        writer = new(path);
        bitMerger = new();
    }

    private BufferedWriter writer;
    private BitMerger bitMerger;

    public async Task WriteBitAsync(Bit value)
    {
        bitMerger.PushBit(value);
        if (bitMerger.ByteReady)
        {
            await writer.Write(bitMerger.GetByte());
        }
    }

    public async Task WriteByteAsync(byte value)
    {
        bitMerger.PushByte(value);
        if (bitMerger.ByteReady)
        {
            var b = bitMerger.GetByte();
            await writer.Write(b);
        }
        await writer.Flush();
    }

    public async Task WriteIntAsync(int value)
    {
        var b = (byte)(value >>> 24);
        await writer.Write(b);
        b = (byte)(value >>> 16);
        await writer.Write(b);
        b = (byte)(value >>> 8);
        await writer.Write(b);
        b = (byte)value;
        await writer.Write(b);
    }

    public async Task<byte> CloseAsync()
    {
        byte pbc = 0;
        if (bitMerger.ByteInProgress)
        {
            while (!bitMerger.ByteReady)
            {
                pbc++;
                bitMerger.PushBit(new());
            }
            await writer.Write(bitMerger.GetByte());
        }
        await writer.CloseAsync();
        return pbc;
    }
}
